### ISSUE TITLE
[FR] configure e-steps directly from printer

### ISSUE THREAD

**Issue by ratwix** [CR6Community/Marlin#57]
### Description

Add an option to configure automaticaly e-steps in the printer without Octoprint or a USB cable plugged to the PC.

### Feature Workflow

1. Select in the menu somewhere "configure e-steps"
2. Follow the instruction (like [here](https://teachingtechyt.github.io/calibration.html#esteps)) : mark your finament @120mm, run the procedure, measure agains the mark, enter the value
3. Printer compute & save e-steps

#### Additional Information

* CR6-SE USB port is crap, if we can avoid using it, it will save a lot of USB ports & motherboards

**Comment by RandyRVA** [comment]
Yes, this would be a great feature.  I killed my computer and printer by connecting the usb cable to try to do this.

**Comment by dryphi** [comment]
I agree this could be helpful esp given the USB problems on the CR6.
In the meantime, this can be done from an SD card by adding commands to the start G-code.
First mark the material at a known distance from the extruder (usually 110 or 120mm). Then put the following commands into the start G-code of an otherwise blank print:
G1 E100 F100
--> Determine how much actually extrudes (mark the material beforehand)
--> Enew = Eold*100/(length actually extruded). Default E-step value from the factory is 93 I believe. So this formula becomes Enew = 9300 / (Length extruded).

With a new "blank" print, add the following commands to the start G-code instead:
M92 E###.# (enter the new value)
M500

**Comment by RandyRVA** [comment]
Yes but there is no way to know what the current esteps value is without a terminal window...  I think?

**Comment by dryphi** [comment]
> Yes but there is no way to know what the current esteps value is without a terminal window... I think?

As far as I know that's correct. The M503 command outputs to the terminal but not to the LCD screen. Hence my suggestion is to base it on the factory value (93 I think). You can set it to 93 (or any value) before you start using M92 E93 then do the extrusion. That way you know exactly the value you're starting with.
Also you can adjust the extrusion multiplier (flow rate in Cura). This does essentially the same thing - just multiplies the E-steps by whatever percentage you input. When you're happy with how it extrudes calculate the value and enter it using the M92 command from an SD card. Be sure to use M500 to save it to EEPROM.

**Comment by Sebazzz** [comment]
This is a relatively low priority feature imho though. You only do it once, then leave it.

(Don't forget that changing e-steps is often hiding your problem of incorrect tension on the feeder. The gear ratio is fixed, the number of steps the stepper has to make for a turn too.)

**Comment by ratwix** [comment]
> I agree this could be helpful esp given the USB problems on the CR6.
> In the meantime, this can be done from an SD card by adding commands to the start G-code.
> First mark the material at a known distance from the extruder (usually 110 or 120mm). Then put the following commands into the start G-code of an otherwise blank print:
> G1 E100 F100
> --> Determine how much actually extrudes (mark the material beforehand)
> --> Enew = Eold*100/(length actually extruded). Default E-step value from the factory is 93 I believe. So this formula becomes Enew = 9300 / (Length extruded).
> 
> With a new "blank" print, add the following commands to the start G-code instead:
> M92 E###.# (enter the new value)
> M500

You need to know your "old" esteps, and for that, you need to plus the USB cable. So no, you can not be 100% independent with SD CARD

**Comment by Sebazzz** [comment]
The default e-steps is 93 - besides you can set it to a known value and calibrate from there.

**Comment by Allan-N** [comment]
Question.  Is the default e-steps value (93) one that just happens to work well for most people or was it computed based on the stepper motor and gear ratio?

**Comment by dryphi** [comment]
> Question. Is the default e-steps value (93) one that just happens to work well for most people or was it computed based on the stepper motor and gear ratio?

That's the Creality default. I would assume it's based on the extruder gear diameter.

**Comment by dryphi** [comment]
> You need to know your "old" esteps, and for that, you need to plus the USB cable. So no, you can not be 100% independent with SD CARD

Yes you can. Just set it to a known value such (e.g. 93) using M92 E93 then go from there.

**Comment by Sebazzz** [comment]
> 
> 
> Question. Is the default e-steps value (93) one that just happens to work well for most people or was it computed based on the stepper motor and gear ratio?

Yes. You shouldn't need to change the esteps with the default extruder. Instead, tension the tensioning screw.

**Comment by nicholaspearson** [comment]
> > Question. Is the default e-steps value (93) one that just happens to work well for most people or was it computed based on the stepper motor and gear ratio?
> 
> Yes. You shouldn't need to change the esteps with the default extruder. Instead, tension the tensioning screw.

I disagree, the E-Steps on almost every machine I've ever touched was different than 93.

This should be configurable from the touch interface, the CR-20 Pro 3DPC firmware allows for this and it's a life saver. So much time saved.

Default it to the current E-Steps and then allow the user to change it from there, this is 3D Printing core functionality.

**Comment by dryphi** [comment]
It's a good idea. Try posting to Marlin Main as this would affect all printers running Marlin in addition to this small branch.

**Comment by Thinkersbluff** [comment]
I already have the ability to enter parameters like esteps and driver current directly through the LCD menus on my Ender 3, so Marlin may support the functionality, but only on an LCD controller?

NOTE: Don't know whether it is just me (not a programmer), but VSCode shows "HAS_TRINAMIC_CONFIG" as greyed-out in my extui 30Nov firmware Configuration_adv.h, but it is NOT greyed-out in my Marlin 4.0.7.2 config for the Ender 3 (with BTT_SKR_MINI_E3.)

**Comment by Sebazzz** [comment]
It is supported but not *through* the LCD. Only though gcode or host programming.

Met vriendelijke groet,
Sebastiaan Dammann
________________________________
Van: Thinkersbluff <notifications@github.com>
Verzonden: Saturday, December 12, 2020 6:00:15 PM
Aan: CR6Community/Marlin <Marlin@noreply.github.com>
CC: Sebastiaan Dammann <sebastiaandammann@outlook.com>; Comment <comment@noreply.github.com>
Onderwerp: Re: [CR6Community/Marlin] [FR] configure e-steps directly from printer (#57)


I already have the ability to enter parameters like esteps and driver current directly through the LCD menus on my Ender 3, so Marlin may support the functionality, but only on an LCD controller?

**Comment by dryphi** [comment]
Suggest we merge this with https://github.com/CR6Community/Marlin/issues/60 and https://github.com/CR6Community/Marlin/issues/66

**Comment by Dreambeemer** [comment]
It would be realy nice when you could change the e-steps from the screen without using the USB-port ore changing the G-codes.
I'am a totally noob in 3D-printing, so for me it should be a big releave. I'am very afraid to mess around with the G-codes, because I don't know how it works. Release 5 works like a charm. Credits for you all who worked it out.

**Comment by Sebazzz** [comment]
Current working concept - ignore the user interface work for now, that's not my specialty, this is just getting the "back-end" ready:

**Heat up to target temperature**
![image](https://user-images.githubusercontent.com/1426097/103481983-7f5fdf00-4dde-11eb-8a91-073227913ac3.png)

**Extruding**
![image](https://user-images.githubusercontent.com/1426097/103481988-82f36600-4dde-11eb-9642-7193d3d94713.png)

**Cue to measure remaining filament**
![image](https://user-images.githubusercontent.com/1426097/103482007-8dadfb00-4dde-11eb-90b6-a4da3923ac34.png)

**New e-steps are calculated for you**
![image](https://user-images.githubusercontent.com/1426097/103482014-9d2d4400-4dde-11eb-9225-c955cb6a55e0.png)

**And apply them if desired**
![image](https://user-images.githubusercontent.com/1426097/103482017-a61e1580-4dde-11eb-8530-164c5272fcea.png)

You can also directly enter a value in the "your new e-steps" field and apply if you don't like being guided.

**Comment by Dreambeemer** [comment]
Waauw, that’s quick. I can’t wait to use it.
Great job👍🏻 Where should we be without you guys.

**Comment by ratwix** [comment]
excellent ! thanks

**Comment by JamesSF69** [comment]
Yes, please!!! Thank you for getting this on the agenda! Hope to use it soon!

**Comment by Sebazzz** [comment]
In [#prelease-testing channel](https://discord.com/channels/759603746270609428/791082544899555358/797579692747784232) of our Discord an early build is available if anyone wants to try ([invite link if you haven't joined the Discord earlier](https://discord.gg/RKrxYy3Q9N)).

**Comment by Thinkersbluff** [comment]
I tried the new esteps function with the “free air” method described in Luke’s hot end fix document. I have his fix installed. I use Capricorn tubing, an all-metal extruder cold end with the stock extruder gear and an otherwise stock CR-6 SE hot end and 0.4mm stock nozzle.
I removed the Bowden tube from the cold end of the extruder and measured the filament extruded from or retracted into the cold end outlet.

Marlin estimates the esteps for this extruder cold end to be 93, on the Ender3, from whence I took it.  It works with precisely the same gear as fitted in the stock C-6 SE extruder.

The extrude test calculated 98.7 steps if I extruded into the Bowden tube, but 95.1 if I extruded into free air. I think some advice may be necessary on how this test is meant to be performed, to help Users avoid inadvertently creating themselves a problem calibrating their extruders with friction in the system.

I also tried with and without the filament runout sensor in series with the extruder, just to be sure it did not impact the extruder’s performance. I saw no impact on the esteps..

Overall, I liked the interface. If possible, it would help to give feedback like “E-Heating” while the nozzle is heating. I was confused at first, when it said “Extruding” but it wasn’t.
I don’t like s/mm as units, since s already means seconds.  I recommend either spell it out in full or put the units in the label and just write the value.

I did notice that the esteps function extrude rate is much slower than is programmed for the load/unload function. I thought it much too slow. Is there a design reason for that rate?

**Comment by Thinkersbluff** [comment]
NOTE: We "Brits" have a saying, "Much wants more."

While I was enjoying this new powerful support feature for eSteps calibration, I thought, "Hey! This would be really easy to adapt into a %Flow calibration tool!"

**Comment by Sebazzz** [comment]
> The extrude test calculated 98.7 steps if I extruded into the Bowden tube, but 95.1 if I extruded into free air. I think some advice may be necessary on how this test is meant to be performed, to help Users avoid inadvertently creating themselves a problem calibrating their extruders with friction in the system.

I'm not sure where we have a place for such guidance or if that is even up (in scope) for the firmware to do. Unless DWIN has something magical to easily adding such guidance, this would require adding pages.  That said, opinions differ on whether to use a free air method or not. These screens are obviously tailored for the latter (but also usable for the former).

> I don’t like s/mm as units, since s already means seconds. I recommend either spell it out in full or put the units in the label and just write the value.

I've removed the units completely (note: took the screenshots before I did that).

> I did notice that the esteps function extrude rate is much slower than is programmed for the load/unload function. I thought it much too slow. Is there a design reason for that rate?

I'm using the recommended speed from the Teaching Tech website.

> While I was enjoying this new powerful support feature for eSteps calibration, I thought, "Hey! This would be really easy to adapt into a %Flow calibration tool!"

You're keeping us busy 😉 I'm not sure if we want to implement a full calibration suite here. You'd still need to slice a cube in vase mode to do the actual test, and there is also not anything to save at the Marlin side.

----

Okay, current state of things:
- @grobux cleaned up my pink comic sans into something nice
- I did some changes on the Marlin side for the changes at the touch screen side

This is how it now looks:

![20210126_204048187_iOS](https://user-images.githubusercontent.com/1426097/105904329-4b648c00-6021-11eb-8ff9-1e960fd15252.jpg)

![20210126_204206126_iOS](https://user-images.githubusercontent.com/1426097/105904352-53243080-6021-11eb-94cc-0cc63ed29a5f.jpg)

![20210126_204343672_iOS](https://user-images.githubusercontent.com/1426097/105904357-54edf400-6021-11eb-9ef9-81f3b00474c2.jpg)

With this I consider the feature implemented. Keep watching the #prerelease-testing channel on the Discord for a build if you want to play with it.
